src/datetime.rs

```diff
@@ -0,0 +1,218 @@
+use std::{
+    fmt::{self, Display},
+    time::{Duration, SystemTime},
+};
+
+use chrono::{LocalResult, TimeZone, Utc};
+
+/// Struct representing a BSON datetime.
+/// Note: BSON datetimes have millisecond precision.
+///
+/// To enable conversions between this type and [`chrono::DateTime`], enable the `"chrono-0_4"`
+/// feature flag in your `Cargo.toml`.
+/// ```
+/// use chrono::prelude::*;
+/// # fn main() -> std::result::Result<(), Box<dyn std::error::Error>> {
+/// # #[cfg(feature = "chrono-0_4")]
+/// # {
+/// let chrono_dt: chrono::DateTime<Utc> = "2014-11-28T12:00:09Z".parse()?;
+/// let bson_dt: bson::DateTime = chrono_dt.into();
+/// let bson_dt = bson::DateTime::from_chrono(chrono_dt);
+/// let back_to_chrono: chrono::DateTime<Utc> = bson_dt.into();
+/// let back_to_chrono = bson_dt.to_chrono();
+/// # }
+/// # Ok(())
+/// # }
+/// ```
+///
+/// This type differs from [`chrono::DateTime`] in that it serializes to and deserializes from a
+/// BSON datetime rather than an RFC 3339 formatted string. Additionally, in non-BSON formats, it
+/// will serialize to and deserialize from that format's equivalent of the
+/// [extended JSON representation](https://docs.mongodb.com/manual/reference/mongodb-extended-json/) of a datetime.
+/// To serialize a [`chrono::DateTime`] as a BSON datetime, you can use
+/// [`serde_helpers::chrono_datetime_as_bson_datetime`].
+///
+/// ```rust
+/// # #[cfg(feature = "chrono-0_4")]
+/// # {
+/// use serde::{Serialize, Deserialize};
+///
+/// #[derive(Serialize, Deserialize)]
+/// struct Foo {
+///     // serializes as a BSON datetime.
+///     date_time: bson::DateTime,
+///
+///     // serializes as an RFC 3339 / ISO-8601 string.
+///     chrono_datetime: chrono::DateTime<chrono::Utc>,
+///
+///     // serializes as a BSON datetime.
+///     // this requires the "chrono-0_4" feature flag
+///     #[serde(with = "bson::serde_helpers::chrono_datetime_as_bson_datetime")]
+///     chrono_as_bson: chrono::DateTime<chrono::Utc>,
+/// }
+/// # }
+/// ```
+#[derive(Eq, PartialEq, Ord, PartialOrd, Hash, Copy, Clone)]
+pub struct DateTime(i64);
+
+impl crate::DateTime {
+    /// The latest possibile date that can be represented in BSON.
+    pub const MAX: crate::DateTime = DateTime(i64::MAX);
+
+    /// The earliest possibile date that can be represented in BSON.
+    pub const MIN: crate::DateTime = DateTime(i64::MIN);
+
+    /// Makes a new [`DateTime`] from the number of non-leap milliseconds since
+    /// January 1, 1970 0:00:00 UTC (aka "UNIX timestamp").
+    pub fn from_millis(date: i64) -> Self {
```

Comment:
I thought about updating this to be `const`, but I was hesitant to do so because `chrono::DateTime` lacks a `const` constructor, and in case `chrono` ever did start supporting the entire range of possible BSON datetimes in the future, we would want to go back to storing one internally. If this were `const`, we'd be pretty much stuck with an internal `i64` representation forever though.

All that being said, it seems extremely unlikely `chrono` will change the range of dates they support to the BSON range, so it's probably best to just go with `const` here. Updated.

cc @isabelatkinson @kmahar